### ISSUE TITLE
using addStream with never ending observable on Broadcast stream can't be canceled

### ISSUE THREAD

**Issue by audkar** [dart-lang/sdk#38034]
Example:
```dart
final controller = StreamController.broadcast();

Observable.periodic(Duration(seconds: 1))
  .doOnData((_) => print('tick'))
  .pipe(controller.sink);

  controller.stream.listen((_) {}).cancel();
  // controller.close(); <-- This can't be called because "ERROR: Bad state: Cannot add new events while doing an addStream"
```

`tick` will be printed forever. 

When BroadcastStream is changed to simple Stream when cancellation is propagated correctly.

*Dart version 2.4.0*

**Comment by lrhn** [comment]
This is expected behavior.

A non-broadcast stream controller shuts down when the listener cancels. This cancels the `addStream` as well, which you can see from the future returned by `addStream` completing. At that point, the controller is in a state with no listeners and no added streams. Calling `close` is irrelevant because nobody will ever see the event anyway, it's just a no-op (apart from creating the `done` future).

A broadcast stream controller stays active when a listener cancels. There may be more listeners already, or they might be added later. So, it also does not cancel the current `addStream`, meaning that the `addStream` future doesn't complete, and you can't call `close`.

**Comment by audkar** [comment]
I understand your explanation from implementation side but I still can't get it from API usability. 

Can you confirm that API is designed in a way that:  **executing particular code creates scenario where resource is leaked without possibility to remove it**?

I am using `Stream.periodic` for simplicity. But in real life this can be 3rd party library stream which holds a lot of resources. Using such stream with with `addStream` method will make that stream non-cancelable and resources leak permanently. Do I understand correctly this intended behavior?

**Comment by lrhn** [comment]
Yes. 

It is intended behavior for a broadcast stream controller to keep running when nobody is listening.

Adding a never-ending stream to a broadcast stream controller will create a never-ending broadcast stream unless the owner of the controller does something to prevent it.

The producer of a broadcast stream is supposed to react to situations where there are no listeners *if they want to*, through the `onListen` and `onCancel` futures. However, the `addStream` functionality does not give you a way to interrupt it. If you need to do that, you should be adding the individual events manually instead of using `addStream`, or wrapping the added stream in a way that allows you to interrupt it.

There are many ways to create code which runs forever, like loops. They all require the author to take some step to avoid running forever. The main difference between this and a `for (;;);` is that the latter will *block*, but persistent asynchronous code will allow other code to run as well, so you won't recognize it as soon. 

If you write:
```dart
foo(int x) async {
  for (int y = 0; y != x; y++) {
    await something(y);
  }
}
...
 foo(null);
```
then your code will also run forever with no way to stop it.

So, in short, do not add an infinite stream to a broadcast stream controller because that *will* create an infinite broadcast stream with no way to stop it.

**Comment by audkar** [comment]
Thank you for such detailed explanation :bowing_man: 

Just small nitpicking - plain `Future` can't be compared with `Stream`. `Stream` exposes contract that it can be canceled. Problem is that using `addStream` with broadcast removes it. Fair comparison would be comparing it with `CancelableOperation`.

So comparison should be:
```dart
foo(int x) => CancelableOperation.fromFuture(() async {
  for (int y = 0; y != x; y++) {
    await something(y);
  }
})
...
foo(null)
  .then((_){}) <--- this is equivalent for addStream with broadcast
  .cancel();
```
And `CancelableOperation` with `then` provides an [option](https://pub.dev/documentation/async/latest/async/CancelableOperation/then.html) to propagate cancelation. 
```dart
foo(null)
  .then((_){}, propagateCancel: true)
  .cancel();
```

> the addStream functionality on a broadcast controller might be the issue here.  It does not allow you to stop it again, from either end.

Yes, this is issue I am seeing. Maybe I didn't expressed myself clearly :) I don't know dart API good enough to suggest reasonable solution. But I would expect such option. Maybe smth like:
```dart
broadcastStream.close({cancelAddedStream: false});
```

**Comment by lrhn** [comment]
I have a hope to change stream controller so that you can add more than one stream at a time, and add other events while adding a stream. It's not that controversial from an API perspective, but it does require some rewriting of the underlying classes.
If we do that, then calling `close` should also cancel any current active `addStream` (since no further events would be allowed anyway).

That's still a little heavy-handed. If you just want to stop the `addStream` without closing everything, you still need another handle to do that.
